File: app/src/main/java/io/github/freewebmovement/igniter/exempt/contract/ExemptAppContract.java
package io.github.freewebmovement.igniter.exempt.contract;

import androidx.annotation.AnyThread;
import androidx.annotation.UiThread;

import java.util.List;

import io.github.freewebmovement.igniter.common.mvp.BasePresenter;
import io.github.freewebmovement.igniter.common.mvp.BaseView;
import io.github.freewebmovement.igniter.exempt.data.AppInfo;

public interface ExemptAppContract {
    interface Presenter extends BasePresenter {
        void updateAppInfo(AppInfo appInfo, int position, boolean exempt);

        void saveExemptAppInfoList();

        /**
         * @return true if exit directly, false to cancel exiting.
         */
        boolean handleBackPressed();

        void filterAppsByName(String name);

        void exit();
    }

    interface View extends BaseView<Presenter> {
        @UiThread
        void showLoading();

        @UiThread
        void dismissLoading();

        @UiThread
        void showSaveSuccess();

        @UiThread
        void showExitConfirm();

        @UiThread
        void showAppList(List<AppInfo> packageNames);

        @AnyThread
        void exit(boolean configurationChanged);
    }
}

File: app/src/main/java/io/github/freewebmovement/igniter/activities/exempt/contract/ExemptAppContract.kt
package io.github.freewebmovement.igniter.activities.exempt.contract

import androidx.annotation.AnyThread
import androidx.annotation.UiThread
import io.github.freewebmovement.igniter.common.mvp.BasePresenter
import io.github.freewebmovement.igniter.common.mvp.BaseView
import io.github.freewebmovement.igniter.persistence.data.AppInfo

interface ExemptAppContract {
    interface Presenter : BasePresenter {
        fun updateAppInfo(appInfo: AppInfo, position: Int, exempt: Boolean)

        fun saveExemptAppInfoList()

        /**
         * @return true if exit directly, false to cancel exiting.
         */
        fun handleBackPressed(): Boolean

        fun filterAppsByName(name: String)

        fun exit()
    }

    interface View : BaseView<Presenter> {
        @UiThread
        fun showLoading()

        @UiThread
        fun dismissLoading()

        @UiThread
        fun showSaveSuccess()

        @UiThread
        fun showExitConfirm()

        @UiThread
        fun showAppList(packageNames: List<AppInfo>)

        @AnyThread
        fun exit(configurationChanged: Boolean)
    }
}
